File: typescript/src/solver.ts
/*
 * Copyright (C) 2007 Time Intermedia Corporation <puzzle@timedia.co.jp>
 * TypeScript rewrite derived from NPGenerator V2.0.2.
 * Director: Hirofumi Fujiwara / Puzzler: Naoki Inaba / Programmer: Masaya Kiwada
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

import { bit, KindOfAnswer, SolverMethod, Status } from "./core.js";

export function swordfish(state: Status): boolean {
  let updated = false;
  const size = state.getSize();
  const pairs = new Int32Array(size * 2);
  const bitCounts = state.block.bitCounts;
  for (let number = 1; number <= size; number++) {
    for (let block = 0; block < size * 2; block++) {
      pairs[block] = state.getCandCountOfBlock(block, number) > 3
        ? -1
        : state.getCandidatePositionMask(block, number);
    }
    for (let first = 0; first < size; first++) {
      if (pairs[first]! <= 0) continue;
      for (let second = first + 1; second < size; second++) {
        if (pairs[second]! <= 0) continue;
        for (let third = second + 1; third < size; third++) {
          if (pairs[third]! <= 0) continue;
          let union = pairs[first]! | pairs[second]! | pairs[third]!;
          if (
            (bitCounts === undefined ? bit.numberOf1Bits(union) : bitCounts[union]!) !== 3
          ) continue;
          const excluded = (1 << first) | (1 << second) | (1 << third);
          while (union !== 0) {
            const position = 31 - Math.clz32(union);
            const group = position + size;
            const cells = state.getBlock(group);
            let targets = state.getCandidatePositionMask(group, number) & ~excluded;
            while (targets !== 0) {
              const targetBit = targets & -targets;
              if (removeCand(state, cells[bit.ntz(targetBit)]!, number)) updated = true;
              targets ^= targetBit;
            }
            union &= ~(1 << position);
          }
        }
      }
    }
    for (let first = size; first < size * 2; first++) {
      if (pairs[first]! <= 0) continue;
      for (let second = first + 1; second < size * 2; second++) {
        if (pairs[second]! <= 0) continue;
        for (let third = second + 1; third < size * 2; third++) {
          if (pairs[third]! <= 0) continue;
          let union = pairs[first]! | pairs[second]! | pairs[third]!;
          if (
            (bitCounts === undefined ? bit.numberOf1Bits(union) : bitCounts[union]!) !== 3
          ) continue;
          const excluded =
            (1 << (first - size)) |
            (1 << (second - size)) |
            (1 << (third - size));
          while (union !== 0) {
            const position = 31 - Math.clz32(union);
            const cells = state.getBlock(position);
            let targets = state.getCandidatePositionMask(position, number) & ~excluded;
            while (targets !== 0) {
              const targetBit = targets & -targets;
              if (removeCand(state, cells[bit.ntz(targetBit)]!, number)) updated = true;
              targets ^= targetBit;
            }
            union &= ~(1 << position);
          }
        }
      }
    }
  }
  return updated;
}

export function xWing(state: Status): boolean {
  let updated = false;
  const size = state.getSize();
  const divisor = size * size + 1;
  const pairs = new Int32Array(size * 2);
  const rows = state.block.rowOfCell;
  const columns = state.block.columnOfCell;
  for (let number = 1; number <= size; number++) {
    for (let block = 0; block < size * 2; block++) {
      if (state.getCandCountOfBlock(block, number) !== 2) {
        pairs[block] = -1;
        continue;
      }
      let positions = state.getCandidatePositionMask(block, number);
      const firstPosition = bit.ntz(positions);
      positions &= positions - 1;
      const secondPosition = bit.ntz(positions);
      const cells = state.getBlock(block);
      pairs[block] = cells[firstPosition]! * divisor + cells[secondPosition]!;
    }
    for (let first = 0; first < size; first++) {
      if (pairs[first]! <= 0) continue;
      for (let second = first + 1; second < size; second++) {
        if (pairs[second]! <= 0) continue;
        const a1 = (pairs[first]! / divisor) | 0;
        const a2 = pairs[first]! % divisor;
        const b1 = (pairs[second]! / divisor) | 0;
        const b2 = pairs[second]! % divisor;
        const row1 = rows[a1]!;
        const row2 = rows[a2]!;
        if (row1 !== rows[b1] || row2 !== rows[b2]) continue;
        let cells = state.getBlock(row1 + size);
        for (let index = 0; index < cells.length; index++) {
          const cell = cells[index]!;
          if (cell !== a1 && cell !== b1 && removeCand(state, cell, number)) updated = true;
        }
        cells = state.getBlock(row2 + size);
        for (let index = 0; index < cells.length; index++) {
          const cell = cells[index]!;
          if (cell !== a2 && cell !== b2 && removeCand(state, cell, number)) updated = true;
        }
      }
    }
    for (let first = size; first < size * 2; first++) {
      if (pairs[first]! <= 0) continue;
      for (let second = first + 1; second < size * 2; second++) {
        if (pairs[second]! <= 0) continue;
        const a1 = (pairs[first]! / divisor) | 0;
        const a2 = pairs[first]! % divisor;
        const b1 = (pairs[second]! / divisor) | 0;
        const b2 = pairs[second]! % divisor;
        const column1 = columns[a1]!;
        const column2 = columns[a2]!;
        if (column1 !== columns[b1] || column2 !== columns[b2]) continue;
        let cells = state.getBlock(column1);
        for (let index = 0; index < cells.length; index++) {
          const cell = cells[index]!;
          if (cell !== a1 && cell !== b1 && removeCand(state, cell, number)) updated = true;
        }
        cells = state.getBlock(column2);
        for (let index = 0; index < cells.length; index++) {
          const cell = cells[index]!;
          if (cell !== a2 && cell !== b2 && removeCand(state, cell, number)) updated = true;
        }
      }
    }
  }
  return updated;
}

export function nakedTriple(state: Status): boolean {
  let updated = false;
  const size = state.getSize();
  const ids = new Int32Array(size);
  const masks = new Int32Array(size);
  const bitCounts = state.block.bitCounts;
  for (let block = 0; block < state.getBlockNum(); block++) {
    const blockCells = state.getBlock(block);
    let count = 0;
    for (let index = 0; index < blockCells.length; index++) {
      const cell = blockCells[index]!;
      if (state.getCandCountOfCell(cell) <= 3) {
        ids[count] = cell;
        masks[count] = state.getCandidateMask(cell) << 1;
        count++;
      }
    }
    for (let first = 0; first < count; first++) {
      for (let second = 0; second < first; second++) {
        for (let third = 0; third < second; third++) {
          const union = masks[first]! | masks[second]! | masks[third]!;
          if (
            (bitCounts === undefined ? bit.numberOf1Bits(union) : bitCounts[union >> 1]!) !== 3
          ) continue;
          for (let cellIndex = 0; cellIndex < blockCells.length; cellIndex++) {
            const cell = blockCells[cellIndex]!;
            if (cell === ids[first] || cell === ids[second] || cell === ids[third]) continue;
            let values = union & (state.getCandidateMask(cell) << 1);
            while (values !== 0) {
              const number = 31 - Math.clz32(values);
              if (removeCand(state, cell, number)) updated = true;
              values &= ~(1 << number);
            }
          }
        }
      }
    }
  }
  return updated;
}

export function hiddenTriple(state: Status): boolean {
  let updated = false;
  const size = state.getSize();
  const ids = new Int32Array(size);
  const masks = new Int32Array(size);
  const bitCounts = state.block.bitCounts;
  for (let block = 0; block < state.getBlockNum(); block++) {
    const blockCells = state.getBlock(block);
    let count = 0;
    for (let number = 1; number <= size; number++) {
      if (state.getCandCountOfBlock(block, number) <= 3) {
        ids[count] = number;
        masks[count] = state.getCandidatePositionMask(block, number);
        count++;
      }
    }
    for (let first = 0; first < count; first++) {
      for (let second = 0; second < first; second++) {
        for (let third = 0; third < second; third++) {
          let positions = masks[first]! | masks[second]! | masks[third]!;
          if (
            (bitCounts === undefined ? bit.numberOf1Bits(positions) : bitCounts[positions]!) !== 3
          ) continue;
          while (positions !== 0) {
            const position = 31 - Math.clz32(positions);
            const cell = blockCells[position]!;
            if (state.getCandCountOfCell(cell) > 2) {
              for (let number = 1; number <= size; number++) {
                if (
                  number !== ids[first] &&
                  number !== ids[second] &&
                  number !== ids[third] &&
                  removeCand(state, cell, number)
                ) updated = true;
              }
            }
            positions &= ~(1 << position);
          }
        }
      }
    }
  }
  return updated;
}

export function localization(state: Status): boolean {
  let updated = false;
  const details = state.block.intersectionDetails;
  const size = state.getSize();
  const bitCounts = state.block.bitCounts;
  for (let detailIndex = 0; detailIndex < details.length; detailIndex++) {
    const detail = details[detailIndex]!;
    for (let number = 1; number <= size; number++) {
      const firstPositions = state.getCandidatePositionMask(detail.first, number);
      const secondPositions = state.getCandidatePositionMask(detail.second, number);
      const sharedMask = firstPositions & detail.firstMask;
      const shared = bitCounts === undefined
        ? bit.numberOf1Bits(sharedMask)
        : bitCounts[sharedMask]!;
      if (shared === 0) continue;
      let target = -1;
      let commonMask = 0;
      if (
        state.getCandCountOfBlock(detail.first, number) > shared &&
        state.getCandCountOfBlock(detail.second, number) === shared
      ) {
        target = detail.first;
        commonMask = detail.firstMask;
      } else if (
        state.getCandCountOfBlock(detail.first, number) === shared &&
        state.getCandCountOfBlock(detail.second, number) > shared
      ) {
        target = detail.second;
        commonMask = detail.secondMask;
      }
      if (target < 0) continue;
      let targets = state.getCandidatePositionMask(target, number) & ~commonMask;
      const cells = state.getBlock(target);
      while (targets !== 0) {
        const targetBit = targets & -targets;
        if (removeCand(state, cells[bit.ntz(targetBit)]!, number)) updated = true;
        targets ^= targetBit;
      }
    }
  }
  return updated;
}

export function nakedPair(state: Status): boolean {
  let updated = false;
  const size = state.getSize();
  const divisor = size + 1;
  const ids = new Int32Array(size);
  const pairs = new Int32Array(size);
  for (let block = 0; block < state.getBlockNum(); block++) {
    const blockCells = state.getBlock(block);
    let count = 0;
    for (let index = 0; index < blockCells.length; index++) {
      const cell = blockCells[index]!;
      if (state.getCandCountOfCell(cell) !== 2) continue;
      let candidates = state.getCandidateMask(cell);
      const first = bit.ntz(candidates) + 1;
      candidates &= candidates - 1;
      const second = bit.ntz(candidates) + 1;
      ids[count] = cell;
      pairs[count] = first * divisor + second;
      count++;
    }
    for (let first = 0; first < count; first++) {
      for (let second = 0; second < first; second++) {
        if (pairs[first] !== pairs[second]) continue;
        const number1 = (pairs[first]! / divisor) | 0;
        const number2 = pairs[first]! % divisor;
        for (let index = 0; index < blockCells.length; index++) {
          const cell = blockCells[index]!;
          if (cell === ids[first] || cell === ids[second]) continue;
          if (removeCand(state, cell, number1)) updated = true;
          if (removeCand(state, cell, number2)) updated = true;
        }
      }
    }
  }
  return updated;
}

export function hiddenPair(state: Status): boolean {
  let updated = false;
  const size = state.getSize();
  const ids = new Int32Array(size);
  const firstCells = new Int32Array(size);
  const secondCells = new Int32Array(size);
  for (let block = 0; block < state.getBlockNum(); block++) {
    const blockCells = state.getBlock(block);
    let count = 0;
    for (let number = 1; number <= size; number++) {
      if (state.getCandCountOfBlock(block, number) !== 2) continue;
      let positions = state.getCandidatePositionMask(block, number);
      const firstPosition = bit.ntz(positions);
      positions &= positions - 1;
      const secondPosition = bit.ntz(positions);
      ids[count] = number;
      firstCells[count] = blockCells[firstPosition]!;
      secondCells[count] = blockCells[secondPosition]!;
      count++;
    }
    for (let first = 0; first < count; first++) {
      for (let second = 0; second < first; second++) {
        if (
          firstCells[first] !== firstCells[second] ||
          secondCells[first] !== secondCells[second]
        ) continue;
        let cell = firstCells[first]!;
        for (let cellIndex = 0; cellIndex < 2; cellIndex++) {
          if (state.getCandCountOfCell(cell) > 2) {
            for (let number = 1; number <= size; number++) {
              if (
                number !== ids[first] &&
                number !== ids[second] &&
                removeCand(state, cell, number)
              ) updated = true;
            }
          }
          cell = secondCells[first]!;
        }
      }
    }
  }
  return updated;
}

export function removeCand(state: Status, cell: number, number: number): boolean {
  if (!state.deleteCandidate(cell, number)) return false;
  if (
    state.unique.cellUnique &&
    state.getCandCountOfCell(cell) === 1 &&
    state.isEmptyCell(cell)
  ) {
    addNumber(state, cell, state.getUniqueCandidate(cell));
  }
  const memberships = state.block.getWhere(cell);
  for (let index = 0; index < memberships.length; index++) {
    const block = memberships[index]!;
    if (!state.unique.vhUnique && state.isVHBlock(block)) continue;
    if (!state.unique.blockUnique && !state.isVHBlock(block)) continue;
    const positions = state.getCandidatePositionMask(block, number);
    if (positions !== 0 && (positions & (positions - 1)) === 0) {
      addNumber(state, state.getBlock(block)[bit.ntz(positions)]!, number);
    }
  }
  return true;
}

function deleteCandPeer(state: Status, cell: number, number: number): void {
  const size = state.getSize();
  for (let candidate = 1; candidate <= size; candidate++) {
    if (candidate !== number) removeCand(state, cell, candidate);
  }
  const memberships = state.block.getWhere(cell);
  for (let index = 0; index < memberships.length; index++) {
    const peers = state.getBlock(memberships[index]!);
    for (let peerIndex = 0; peerIndex < peers.length; peerIndex++) {
      const peer = peers[peerIndex]!;
      if (cell !== peer) removeCand(state, peer, number);
    }
  }
}

export function addNumber(state: Status, cell: number, number: number): boolean {
  if (!state.assignValue(cell, number)) return false;
  deleteCandPeer(state, cell, number);
  return true;
}

export function answer(state: Status, method: SolverMethod): Status {
  state.unique = method.unique;
  const cells = state.cells();
  for (let cell = 0; cell < cells.length; cell++) {
    const number = cells[cell]!;
    if (number === 0) continue;
    if (!state.isCand(cell, number)) {
      state.setKindOfAnswer(KindOfAnswer.IRREGULAR_PROBLEM);
    }
    deleteCandPeer(state, cell, number);
  }
  if (state.isInvalid()) return state;
  let updated = true;
  while (updated) {
    updated = false;
    if (state.getSpaceCount() === 0) break;
    if (method.localization && !updated) updated = localization(state);
    if (method.nakedPair && !updated) updated = nakedPair(state);
    if (method.hiddenPair && !updated) updated = hiddenPair(state);
    if (method.XWing && !updated) updated = xWing(state);
    if (method.nakedTriple && !updated) updated = nakedTriple(state);
    if (method.hiddenTriple && !updated) updated = hiddenTriple(state);
    if (method.swordfish && !updated) updated = swordfish(state);
    if (state.isNoAnswer()) return state;
  }
  state.setKindOfAnswer(
    state.getSpaceCount() > 0
      ? KindOfAnswer.MULTIPLE_ANSWER
      : KindOfAnswer.UNIQUE_ANSWER,
  );
  return state;
}

File: rust/src/solver.rs
/*
 * Copyright (C) 2007 Time Intermedia Corporation <puzzle@timedia.co.jp>
 * Rust rewrite derived from NPGenerator V2.0.2.
 *
 * Number Place Generator Version 2.0
 * Director: Hirofumi Fujiwara / Puzzler: Naoki Inaba
 * Programmer: Masaya Kiwada
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

use crate::model::{AnswerKind, SolverMethod, Status};

const MAX_BLOCKS: usize = 50;

struct StackList<T: Copy> {
    values: [T; MAX_BLOCKS],
    len: usize,
}

impl<T: Copy> StackList<T> {
    #[inline]
    fn new(initial: T) -> Self {
        Self {
            values: [initial; MAX_BLOCKS],
            len: 0,
        }
    }

    #[inline]
    fn clear(&mut self) {
        self.len = 0;
    }

    #[inline]
    fn push(&mut self, value: T) {
        self.values[self.len] = value;
        self.len += 1;
    }

    #[inline]
    fn len(&self) -> usize {
        self.len
    }
}

impl<T: Copy> std::ops::Index<usize> for StackList<T> {
    type Output = T;

    #[inline]
    fn index(&self, index: usize) -> &Self::Output {
        &self.values[index]
    }
}

pub fn add_number(state: &mut Status, cell: usize, n: i32) -> bool {
    if state.assign_value(cell, n) {
        delete_cand_peer(state, cell, n);
        true
    } else {
        false
    }
}

fn delete_cand_peer(state: &mut Status, cell: usize, n: i32) {
    for x in 1..=state.size as i32 {
        if x != n {
            remove_cand(state, cell, x);
        }
    }
    let membership_count = state.block.where_cell[cell].len();
    for membership_index in 0..membership_count {
        let block = state.block.where_cell[cell][membership_index];
        let mut positions = state.candidate_index_mask(block, n);
        while positions != 0 {
            let cell_index = positions.trailing_zeros() as usize;
            positions &= positions - 1;
            let peer = state.block.blocks[block][cell_index];
            if peer != cell {
                remove_cand(state, peer, n);
            }
        }
    }
}

fn remove_cand(state: &mut Status, cell: usize, n: i32) -> bool {
    if !state.delete_candidate(cell, n) {
        return false;
    }
    if state.unique.cell_unique && state.is_unique_candidate(cell) && state.cell[cell] == 0 {
        add_number(state, cell, state.unique_candidate(cell));
    }
    let membership_count = state.block.where_cell[cell].len();
    for membership_index in 0..membership_count {
        let block = state.block.where_cell[cell][membership_index];
        if !state.unique.vh_unique && block < state.size * 2 {
            continue;
        }
        if !state.unique.block_unique && block >= state.size * 2 {
            continue;
        }
        let positions = state.candidate_index_mask(block, n);
        if positions != 0 && positions & (positions - 1) == 0 {
            let candidate_cell =
                state.block.blocks[block][positions.trailing_zeros() as usize];
            add_number(state, candidate_cell, n);
        }
    }
    true
}

fn localization(state: &mut Status) -> bool {
    let mut updated = false;
    let mut c1 = [0_u8; 26];
    let mut c2 = [0_u8; 26];
    let mut c12 = [0_u8; 26];
    let intersection_count = state.block.intersection_list.len();
    for intersection_index in 0..intersection_count {
        let intersection = state.block.intersection_list[intersection_index];
        let i = intersection.first;
        let j = intersection.second;
        for n in 1..=state.size {
            c12[n] = (state.candidate_index_mask(i, n as i32)
                & intersection.first_mask)
                .count_ones() as u8;
            c1[n] = state.cand_count_block(i, n as i32) as u8 - c12[n];
            c2[n] = state.cand_count_block(j, n as i32) as u8 - c12[n];
        }
        for n in 1..=state.size {
            if c12[n] == 0 {
                continue;
            }
            if c1[n] > 0 && c2[n] == 0 {
                let mut targets =
                    state.candidate_index_mask(i, n as i32) & !intersection.first_mask;
                while targets != 0 {
                    let cell_index = targets.trailing_zeros() as usize;
                    targets &= targets - 1;
                    let cell = state.block.blocks[i][cell_index];
                    if remove_cand(state, cell, n as i32) {
                        updated = true;
                    }
                }
            } else if c1[n] == 0 && c2[n] > 0 {
                let mut targets =
                    state.candidate_index_mask(j, n as i32) & !intersection.second_mask;
                while targets != 0 {
                    let cell_index = targets.trailing_zeros() as usize;
                    targets &= targets - 1;
                    let cell = state.block.blocks[j][cell_index];
                    if remove_cand(state, cell, n as i32) {
                        updated = true;
                    }
                }
            }
        }
    }
    updated
}

fn naked_pair(state: &mut Status) -> bool {
    let mut updated = false;
    let mut ids = StackList::new(0_usize);
    let mut pairs = StackList::new(0_i32);
    for block in 0..state.block.blocks.len() {
        ids.clear();
        pairs.clear();
        for cell_index in 0..state.block.blocks[block].len() {
            let cell = state.block.blocks[block][cell_index];
            if state.cand_count_cell(cell) == 2 {
                let first_bit = state.cand[cell].trailing_zeros() as i32 + 1;
                let second_bit =
                    (state.cand[cell] & (state.cand[cell] - 1)).trailing_zeros() as i32 + 1;
                ids.push(cell);
                pairs.push(first_bit * (state.size as i32 + 1) + second_bit);
            }
        }
        for j in 0..ids.len() {
            for k in 0..j {
                if pairs[j] == pairs[k] {
                    let p1 = pairs[j] / (state.size as i32 + 1);
                    let p2 = pairs[j] % (state.size as i32 + 1);
                    let cell_count = state.block.blocks[block].len();
                    for cell_index in 0..cell_count {
                        let cell = state.block.blocks[block][cell_index];
                        if cell == ids[j] || cell == ids[k] {
                            continue;
                        }
                        if state.is_cand(cell, p1) {
                            updated |= remove_cand(state, cell, p1);
                        }
                        if state.is_cand(cell, p2) {
                            updated |= remove_cand(state, cell, p2);
                        }
                    }
                }
            }
        }
    }
    updated
}

fn hidden_pair(state: &mut Status) -> bool {
    let mut updated = false;
    let mut ids = StackList::new(0_i32);
    let mut sets = StackList::new(0_u32);
    for block in 0..state.block.blocks.len() {
        ids.clear();
        sets.clear();
        for n in 1..=state.size as i32 {
            let positions = state.candidate_index_mask(block, n);
            if positions.count_ones() == 2 {
                ids.push(n);
                sets.push(positions);
            }
        }
        for i in 0..ids.len() {
            for j in 0..i {
                if sets[i] == sets[j] {
                    let first = sets[i].trailing_zeros() as usize;
                    let second = (sets[i] & (sets[i] - 1)).trailing_zeros() as usize;
                    for &index in &[first, second] {
                        let cell = state.block.blocks[block][index];
                        if state.cand_count_cell(cell) > 2 {
                            for n in 1..=state.size as i32 {
                                if n != ids[i] && n != ids[j] {
                                    updated |= remove_cand(state, cell, n);
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    updated
}

fn naked_triple(state: &mut Status) -> bool {
    let mut updated = false;
    let mut ids = StackList::new(0_usize);
    let mut sets = StackList::new(0_i32);
    for block in 0..state.block.blocks.len() {
        ids.clear();
        sets.clear();
        for cell_index in 0..state.block.blocks[block].len() {
            let cell = state.block.blocks[block][cell_index];
            if state.cand_count_cell(cell) <= 3 {
                ids.push(cell);
                sets.push(state.cand[cell] << 1);
            }
        }
        for j in 0..ids.len() {
            for k in 0..j {
                for l in 0..k {
                    let union: i32 = sets[j] | sets[k] | sets[l];
                    if union.count_ones() != 3 {
                        continue;
                    }
                    let cell_count = state.block.blocks[block].len();
                    for cell_index in 0..cell_count {
                        let cell = state.block.blocks[block][cell_index];
                        if cell == ids[j] || cell == ids[k] || cell == ids[l] {
                            continue;
                        }
                        let mut candidates = union & (state.cand[cell] << 1);
                        while candidates != 0 {
                            let n = 31 - candidates.leading_zeros() as i32;
                            candidates &= !(1 << n);
                            updated |= remove_cand(state, cell, n);
                        }
                    }
                }
            }
        }
    }
    updated
}

fn hidden_triple(state: &mut Status) -> bool {
    let mut updated = false;
    let mut ids = StackList::new(0_i32);
    let mut sets = StackList::new(0_u32);
    for block in 0..state.block.blocks.len() {
        ids.clear();
        sets.clear();
        for n in 1..=state.size as i32 {
            if state.cand_count_block(block, n) <= 3 {
                ids.push(n);
                sets.push(state.candidate_index_mask(block, n));
            }
        }
        for i in 0..ids.len() {
            for j in 0..i {
                for k in 0..j {
                    let union = sets[i] | sets[j] | sets[k];
                    if union.count_ones() != 3 {
                        continue;
                    }
                    for index in (0..state.size).rev() {
                        if union & (1_u32 << index) == 0 {
                            continue;
                        }
                        let cell = state.block.blocks[block][index];
                        if state.cand_count_cell(cell) > 2 {
                            for n in 1..=state.size as i32 {
                                if n != ids[i] && n != ids[j] && n != ids[k] {
                                    updated |= remove_cand(state, cell, n);
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    updated
}

fn xwing(state: &mut Status) -> bool {
    let mut updated = false;
    let size = state.size;
    for n in 1..=size as i32 {
        for i in 0..size {
            let positions = state.candidate_index_mask(i, n);
            if positions.count_ones() != 2 {
                continue;
            }
            for j in i + 1..size {
                if state.candidate_index_mask(j, n) != positions {
                    continue;
                }
                let mut rows = positions;
                while rows != 0 {
                    let row = rows.trailing_zeros() as usize;
                    rows &= rows - 1;
                    let target_block = row + size;
                    let excluded = (1_u32 << i) | (1_u32 << j);
                    let mut targets = state.candidate_index_mask(target_block, n) & !excluded;
                    while targets != 0 {
                        let cell_index = targets.trailing_zeros() as usize;
                        targets &= targets - 1;
                        let cell = state.block.blocks[target_block][cell_index];
                        updated |= remove_cand(state, cell, n);
                    }
                }
            }
        }
        for i in size..size * 2 {
            let positions = state.candidate_index_mask(i, n);
            if positions.count_ones() != 2 {
                continue;
            }
            for j in i + 1..size * 2 {
                if state.candidate_index_mask(j, n) != positions {
                    continue;
                }
                let mut columns = positions;
                while columns != 0 {
                    let column = columns.trailing_zeros() as usize;
                    columns &= columns - 1;
                    let excluded =
                        (1_u32 << (i - size)) | (1_u32 << (j - size));
                    let mut targets = state.candidate_index_mask(column, n) & !excluded;
                    while targets != 0 {
                        let cell_index = targets.trailing_zeros() as usize;
                        targets &= targets - 1;
                        let cell = state.block.blocks[column][cell_index];
                        updated |= remove_cand(state, cell, n);
                    }
                }
            }
        }
    }
    updated
}

fn swordfish(state: &mut Status) -> bool {
    let mut updated = false;
    let size = state.size;
    let mut sets = StackList::new(0_u32);
    for n in 1..=size as i32 {
        sets.clear();
        for block in 0..size * 2 {
            if state.cand_count_block(block, n) > 3 {
                sets.push(0_u32);
            } else {
                sets.push(state.candidate_index_mask(block, n));
            }
        }
        for i in 0..size {
            if sets[i] == 0 {
                continue;
            }
            for j in i + 1..size {
                if sets[j] == 0 {
                    continue;
                }
                for k in j + 1..size {
                    if sets[k] == 0 {
                        continue;
                    }
                    let union = sets[i] | sets[j] | sets[k];
                    if union.count_ones() != 3 {
                        continue;
                    }
                    for group_index in (0..size).rev() {
                        if union & (1_u32 << group_index) == 0 {
                            continue;
                        }
                        let group = group_index + size;
                        let excluded =
                            (1_u32 << i) | (1_u32 << j) | (1_u32 << k);
                        let mut targets =
                            state.candidate_index_mask(group, n) & !excluded;
                        while targets != 0 {
                            let cell_index = targets.trailing_zeros() as usize;
                            targets &= targets - 1;
                            let cell = state.block.blocks[group][cell_index];
                            updated |= remove_cand(state, cell, n);
                        }
                    }
                }
            }
        }
        for i in size..size * 2 {
            if sets[i] == 0 {
                continue;
            }
            for j in i + 1..size * 2 {
                if sets[j] == 0 {
                    continue;
                }
                for k in j + 1..size * 2 {
                    if sets[k] == 0 {
                        continue;
                    }
                    let union = sets[i] | sets[j] | sets[k];
                    if union.count_ones() != 3 {
                        continue;
                    }
                    for group in (0..size).rev() {
                        if union & (1_u32 << group) == 0 {
                            continue;
                        }
                        let excluded = (1_u32 << (i - size))
                            | (1_u32 << (j - size))
                            | (1_u32 << (k - size));
                        let mut targets =
                            state.candidate_index_mask(group, n) & !excluded;
                        while targets != 0 {
                            let cell_index = targets.trailing_zeros() as usize;
                            targets &= targets - 1;
                            let cell = state.block.blocks[group][cell_index];
                            updated |= remove_cand(state, cell, n);
                        }
                    }
                }
            }
        }
    }
    updated
}

pub fn answer_in_place(state: &mut Status, method: SolverMethod) {
    state.unique = method.unique;
    for cell in 0..state.cell.len() {
        let n = state.cell[cell];
        if n != 0 {
            if !state.is_cand(cell, n) {
                state.kind = AnswerKind::Irregular;
            }
            delete_cand_peer(state, cell, n);
        }
    }
    if state.is_invalid() {
        return;
    }
    let mut updated = true;
    while updated {
        updated = false;
        if state.space_count == 0 {
            break;
        }
        if method.localization && !updated {
            updated = localization(state);
        }
        if method.naked_pair && !updated {
            updated = naked_pair(state);
        }
        if method.hidden_pair && !updated {
            updated = hidden_pair(state);
        }
        if method.xwing && !updated {
            updated = xwing(state);
        }
        if method.naked_triple && !updated {
            updated = naked_triple(state);
        }
        if method.hidden_triple && !updated {
            updated = hidden_triple(state);
        }
        if method.swordfish && !updated {
            updated = swordfish(state);
        }
        if state.is_no_answer() {
            return;
        }
    }
    state.kind = if state.space_count > 0 {
        AnswerKind::Multiple
    } else {
        AnswerKind::Unique
    };
}

pub fn answer(mut state: Status, method: SolverMethod) -> Status {
    answer_in_place(&mut state, method);
    state
}
